Sy 2225 improve console data sharing mechanism

pjdotson opened the pull request:
# Issue Pull Request

## Key Information

<!-- Edit the list below with the proper issue number and link -->

- **Linear Issue**: [SY-####](https://linear.app/synnax/issue/)

## Description

<!-- Write a short (2-3 sentence) description describing the changes. -->

## Basic Readiness

- [ ] I have performed a self-review of my code.
- [ ] I have added relevant tests to cover the changes to CI.
- [ ] I have added needed QA steps to the [release candidate](/synnaxlabs/synnax/blob/main/.github/PULL_REQUEST_TEMPLATE/rc.md) template that cover these changes.
- [ ] I have updated in-code documentation to reflect the changes.
- [ ] I have updated user-facing documentation to reflect the changes.

## Backwards Compatibility

### Data Structures

I have ensured that previous versions of stored data structures are properly migrated to new formats in the following projects:

- [ ] Server
- [ ] Console

### API Changes

The following projects have backwards-compatible APIs:

- [ ] Python Client
- [ ] Server
- [ ] TypeScript Client

### Breaking Changes

<!-- If anything in this section is not true, list all breaking changes. -->

emilbon99 commented on client/ts/src/framer/streamer.ts:
```diff
@@ -70,6 +70,14 @@ export class Streamer implements AsyncIterator<Frame>, AsyncIterable<Frame> {
     return this.adapter.adapt(new Frame((await this.stream.receive()).frame));
   }
 
+  updateKeys(keys: channel.Key | channel.Keys): void {
```
See above comment. Why does this function need to exist? Increases API surface area on a critical component of the public client. Is the point that you don't want an async function? I don't see this used anywhere else in the codebase.

emilbon99 commented on client/ts/src/hardware/task/client.ts:
```diff
@@ -411,6 +398,34 @@ export class Client implements AsyncTermSearcher<string, Key, Payload> {
+      task,
+      type,
+      timeout,
+      name ?? (await this.retrieve(task)).name,
```
This function only ever uses `name` if it encounters an error, yet we're doing a `retrieve` without caching on every single call to execute command? Needs to be a better way to handle this without needing to fetch every time.

emilbon99 commented on client/ts/src/hardware/task/client.ts:
```diff
@@ -460,3 +475,73 @@ export class Client implements AsyncTermSearcher<string, Key, Payload> {
+): Promise<string> => {
+  if (frameClient == null) throw NOT_CREATED_ERROR;
+  const key = id.create();
+  await frameClient.write(TimeStamp.now(), COMMAND_CHANNEL_NAME, [
```
Did you test if this actually works? `frameClient.write` uses writer mode persist only by default, so unless there's a bug I don't think this will actually get propagated to devices.

emilbon99 commented on client/ts/src/hardware/task/client.ts:
```diff
@@ -460,3 +475,73 @@ export class Client implements AsyncTermSearcher<string, Key, Payload> {
+  if (frameClient == null) throw NOT_CREATED_ERROR;
+  const streamer = await frameClient.openStreamer(STATE_CHANNEL_NAME);
+  const cmdKey = await executeCommand(frameClient, task, type, args);
+  const timeOut = new TimeSpan(timeout);
```
Wild naming convention. Just name it `parsedTimeout` or something

emilbon99 commented on client/ts/src/hardware/task/client.ts:
```diff
@@ -460,3 +475,73 @@ export class Client implements AsyncTermSearcher<string, Key, Payload> {
+
+const formatTimeoutError = async (
+  type: string,
+  name: string | Promise<string>,
```
I don't think there's anywhere where you actually return a promise here.

emilbon99 replied:
So making this function async is not very useful

emilbon99 commented on client/ts/src/ranger/client.ts:
```diff
@@ -183,10 +186,19 @@ export class Range {
 
-export const sort = (a: Range, b: Range): -1 | 0 | 1 =>
-  sortTimeRange(a.timeRange, b.timeRange);
+  static readonly convertOntologyResourceToPayload = (
```
I don't know I'm kind of skewed on whether it should be a standalone fucntion or not. NOt a big deal, but they should at least be methods and not arrow functions

emilbon99 commented on console/src/channel/useListenForCalculationState.ts:
```diff
@@ -0,0 +1,45 @@
+            description: message,
+            message: `Calculation for ${name} failed`,
+          });
+        } catch {
```
If we're already using `handleError` what's the point of this try catch?

emilbon99 replied:
Or just get rid of the handleError if you want to use the base status.

emilbon99 commented on console/src/range/Explorer.tsx:
```diff
@@ -190,7 +190,7 @@ const ChangeLoader = () => {
   const client = Synnax.use();
   useAsyncEffect(async () => {
-    const obs = await client?.ranges.openTracker();
+    const obs = await client?.ranges.openTracker(); //todo
```
todo?

emilbon99 commented on pluto/src/hardware/task/useCommandExecutor.ts:
```diff
@@ -0,0 +1,99 @@
+      setKey(null);
+    }, timeout.milliseconds);
+  };
+  const addListener = Synch.useAddListener();
```
Any reason why you're not using the `useParsedListener` hook here? Seems like the perfect use case for it

emilbon99 commented on pluto/src/channel/AliasContext.ts:
```diff
@@ -0,0 +1,79 @@
+    const n = await getName(key);
+    setName(n);
+  }, [key, getName]);
+  const rename = useCallback(
```
This seems like the ideal place to use a mutation right? Instead of implementing our own undo logic.

emilbon99 reviewed (changes requested):
This is a good step forward and I think it's close to being able to be merged. That said, there's some things we should still think about:

1. What if I just want a hook called `Channel.use` that always gives me the most up to date version of a channel? I think this is where the power really comes in, because it guarantees that every place we display a channel's properties (such as the vis selector, or edit calculation dialog), we can make sure we always have the most up to date value without needing to keep track of making sure we've set up all of `Synch` listeners.

2. What's the decision on `useMutation` and `useQuery`. There is no consistency between using `useQuery` and then just raw dogging fetches inside of an async effect in other places. There's also some spots where we make modifications and don't use a mutation? Seems like there's a lot of simplification we can do in these places, especially surrounding forms, which is a major pain in the ass rn.

3. As far as a bunch of tiny files goes, I will lodge a small complain against it. There are a bunch of places where a file has 6 lines and is next to 5 or so other logically related files that basically all import from the same place and do very similar things. I don't think that these should be broken up. For example, just put all of the 5 line synchronization hooks into a file called `useSycnchronizers` or `synchronizers`. All for splitting up files when hooks are large and not logically related, but seems excessive in many cases here.

pjdotson commented:
> This is a good step forward and I think it's close to being able to be merged. That said, there's some things we should still think about:
> 
> 1. What if I just want a hook called `Channel.use` that always gives me the most up to date version of a channel? I think this is where the power really comes in, because it guarantees that every place we display a channel's properties (such as the vis selector, or edit calculation dialog), we can make sure we always have the most up to date value without needing to keep track of making sure we've set up all of `Synch` listeners.
> 2. What's the decision on `useMutation` and `useQuery`. There is no consistency between using `useQuery` and then just raw dogging fetches inside of an async effect in other places. There's also some spots where we make modifications and don't use a mutation? Seems like there's a lot of simplification we can do in these places, especially surrounding forms, which is a major pain in the ass rn.
> 3. As far as a bunch of tiny files goes, I will lodge a small complain against it. There are a bunch of places where a file has 6 lines and is next to 5 or so other logically related files that basically all import from the same place and do very similar things. I don't think that these should be broken up. For example, just put all of the 5 line synchronization hooks into a file called `useSycnchronizers` or `synchronizers`. All for splitting up files when hooks are large and not logically related, but seems excessive in many cases here.

1. I added a hook in pluto/ontology called `useResource` that gives an ontology resource for any ontology ID, which should support most of those problems. For specific data types those should be implemented as needed.
2. Decided to mostly not use useQuery, as it isn't super helpful to us unless we seriously wanted to start cacheing server requests. I tried to move most modifications into useMutation hooks, but that might not be 100% yet.
3. changed file sizes in a lot of places

emilbon99 commented on pluto/src/sync/Provider.tsx:
Actually no... I think we should test this now. The point of this CDC refactor is to get our query infrastructure to be production grade. Code that is not tested should, by default, not be considered production grad

emilbon99 commented on pluto/src/hardware/rack/use.ts:
```diff
@@ -0,0 +1,42 @@
+  const handleStateChange = useCallback(
+    (state: rack.State) => {
+      if (rack == null || client == null) return;
+      console.log("state change", state);
```
Remove this log

pjdotson commented on .github/workflows/test.alamos.ts.yaml:
```diff
@@ -12,7 +12,6 @@ on:
       - pnpm-lock.yaml
       - pnpm-workspace.yaml
       - turbo.json
-      - vitest.workspace.ts
```
This config file got removed

pjdotson replied:
I only had it for my editor and then I was having too many problems with VSCode warnings.

pjdotson commented on pluto/src/hooks/ref.ts:
```diff
@@ -47,7 +48,9 @@ export const useStateRef = <T extends state.State>(
 export const useSyncedRef = <T>(value: T): RefObject<T> => {
   const ref = useRef<T>(value);
-  ref.current = value;
+  useEffect(() => {
```
See here under pitfall: https://react.dev/reference/react/useRef#usage

pjdotson commented on pluto/src/hardware/rack/use.ts:
```diff
@@ -0,0 +1,42 @@
+import { useAsyncEffect } from "@/hooks";
+import { Synnax } from "@/synnax";
+
+export const use = (key: rack.Key): rack.Rack | undefined => {
```
At that point we are just reimplementing `useQuery` and should just use that and let it refetch whenever we get a new state. Which I am fine with if you think makes sense, but basically the only useQuery errors we have encountered have been with the task schema parsing in the task dialogs. IMO we should keep data fetching a little split right now between useQuery and useAsyncEffect and later on down the road try to migrate to really take advantage of useQuery's features.
